storage/scylla_api.c

```diff
@@ -0,0 +1,707 @@
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+  }
+
+exit:
+  cass_prepared_free(insert_prepared);
+  return ret;
+}
+
+status_t insert_transaction_into_edgeTable(CassSession* session, scylla_iota_transaction_t* transaction,
+                                           size_t trans_num) {
+  status_t ret = SC_OK;
+  const CassPrepared* insert_prepared = NULL;
+  CassStatement* statement = NULL;
+  const char* insert_query =
+      "INSERT INTO permanent.edgeTable (edge, bundle, address, hash)"
+      "VALUES (?, ?, ?, ?);";
+
+  if (transaction == NULL || trans_num == 0) {
+    return SC_STORAGE_INVAILD_INPUT;
+  }
+
+  if (prepare_query(session, insert_query, &insert_prepared) != CASS_OK) {
+    ta_log_error("%s\n", "prepare INSERT query fail");
+    return SC_STORAGE_CASSANDRA_QUREY_FAIL;
+  }
+
+  for (size_t i = 0; i < trans_num; i++) {
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->address);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->trunk);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+    statement = ret_insert_edge_statement(insert_prepared, transaction + i, (transaction + i)->branch);
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+  }
+
+exit:
+  cass_prepared_free(insert_prepared);
+  return ret;
+}
+
+status_t select_from_bundleTable(CassSession* session, CassStatement* statement,
+                                 scylla_iota_transaction_t** transaction, size_t* transactionNum) {
+  int idx;
+  status_t ret = SC_OK;
+  CassFuture* future = NULL;
+  const CassResult* result;
+  CassIterator* iterator;
+
+  future = cass_session_execute(session, statement);
+  cass_future_wait(future);
+  if (cass_future_error_code(future) != CASS_OK) {
+    print_error(future);
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+    goto exit;
+  }
+
+  /* get transaction number */
+  result = cass_future_get_result(future);
+  iterator = cass_iterator_from_result(result);
+  *transactionNum = 0;
+  while (cass_iterator_next(iterator)) {
+    (*transactionNum)++;
+  }
+  *transaction = malloc((*transactionNum) * sizeof(scylla_iota_transaction_t));
+  if (*transaction == NULL) {
+    ta_log_error("%s\n", "malloc failure");
+    ret = SC_STORAGE_OOM;
+    goto end_iterate;
+  }
+  /* get each transaction result */
+  idx = 0;
+  iterator = cass_iterator_from_result(result);
+  while (cass_iterator_next(iterator)) {
+    const CassRow* row = cass_iterator_get_row(iterator);
+
+    if ((ret = get_blob(row, ((*transaction) + idx)->bundle, "bundle")) != SC_OK) {
+      goto end_iterate;
```

Comment:
why don't you use break?

Reply:
and logger is missing here

Reply:
btw I personally like to move `ret = get_blob(row, ((*transaction) + idx)->address, "address")` out of the if statement. In this way it would easier to set breakpoint and examine the error status code

Reply:
> why don't you use break?

It will lead same result now. Why using break is better here?

Reply:
Doing so, we can remove the label mentioned below.

Reply:
The label used in other checking above.

Reply:
I think if you call `cass_statement_free(statement);` in the first if statement, then you can remove the `end_iterate`